File: ts/src/migrateKey.ts
import { ml_kem768 } from "@noble/post-quantum/ml-kem";
import { ml_dsa65 } from "@noble/post-quantum/ml-dsa";
import { arrayBufferToBase64, base64ToArrayBuffer, keyHash } from "./utils.ts";
import type { EncryptedData, migrateKey, migrateSignKey, Sign } from "./type.ts";

export function generateMigrateKey(): {
  publickKey: string;
  privateKey: string;
} {
  const key = ml_kem768.keygen();
  const publicKeyBinary = arrayBufferToBase64(
    key.publicKey as unknown as ArrayBuffer,
  );
  const privateKeyBinary = arrayBufferToBase64(
    key.secretKey as unknown as ArrayBuffer,
  );
  const publickKey: migrateKey = {
    keyType: "migrateKeyPublic",
    key: publicKeyBinary,
  };
  const privateKey: migrateKey = {
    keyType: "migrateKeyPrivate",
    key: privateKeyBinary,
  };
  return {
    publickKey: JSON.stringify(publickKey),
    privateKey: JSON.stringify(privateKey),
  };
}

export function isValidMigrateKeyPrivate(key: string): boolean {
  if (key.length !== 3240) {
    console.log(key.length);
    return false;
  }
  const { key: keyBinary, keyType } = JSON.parse(key);
  if (keyType !== "migrateKeyPrivate") {
    return false;
  }
  const keyBinaryArray = new Uint8Array(base64ToArrayBuffer(keyBinary));
  if (keyBinaryArray.length !== 2400) {
    console.log(keyBinaryArray.length);
    return false;
  }
  return true;
}

export function isValidMigrateKeyPublic(key: string): boolean {
  if (key.length !== 1619) {
    return false;
  }
  const { key: keyBinary, keyType } = JSON.parse(key);
  if (keyType !== "migrateKeyPublic") {
    return false;
  }
  const keyBinaryArray = new Uint8Array(base64ToArrayBuffer(keyBinary));
  if (keyBinaryArray.length !== 1184) {
    console.log(keyBinaryArray.length);
    return false;
  }
  return true;
}

export async function encryptDataMigrateKey(
  key: string,
  data: string,
): Promise<string | null> {
  if (!isValidMigrateKeyPublic(key)) {
    return null;
  }
  const { key: keyBinary } = JSON.parse(key);
  const dataArray = new TextEncoder().encode(data);
  const ciphertext = ml_kem768.encapsulate(
    new Uint8Array(base64ToArrayBuffer(keyBinary)),
    new Uint8Array(32),
  );
  const ciphertextString = arrayBufferToBase64(
    ciphertext.cipherText as unknown as ArrayBuffer,
  );
  const keyHashString = await keyHash(key);
  const importedKey = await crypto.subtle.importKey(
    "raw",
    new Uint8Array(ciphertext.sharedSecret),
    "AES-GCM",
    true,
    ["encrypt", "decrypt"],
  );
  const iv = crypto.getRandomValues(new Uint8Array(12));
  const encryptedData = await crypto.subtle.encrypt(
    {
      name: "AES-GCM",
      iv: iv,
    },
    importedKey,
    dataArray,
  );
  const viString = arrayBufferToBase64(iv as unknown as ArrayBuffer);
  const encryptedDataString = arrayBufferToBase64(encryptedData);
  const result: EncryptedData = {
    keyType: "migrateKey",
    keyHash: keyHashString,
    encryptedData: encryptedDataString,
    iv: viString,
    cipherText: ciphertextString,
    algorithm: "AES-GCM",
  };
  return JSON.stringify(result);
}

export async function decryptDataMigrateKey(
  key: string,
  data: string,
): Promise<string | null> {
  if (!isValidMigrateKeyPrivate(key)) {
    return null;
  }
  const { key: keyBinary } = JSON.parse(key);
  const { encryptedData, iv, cipherText } = JSON.parse(data);
  if (!cipherText) {
    return null;
  }
  const sharedSecret = ml_kem768.decapsulate(
    new Uint8Array(base64ToArrayBuffer(cipherText)),
    new Uint8Array(base64ToArrayBuffer(keyBinary)),
  );
  const importedKey = await crypto.subtle.importKey(
    "raw",
    new Uint8Array(sharedSecret),
    "AES-GCM",
    true,
    ["encrypt", "decrypt"],
  );
  const decryptedData = await crypto.subtle.decrypt(
    {
      name: "AES-GCM",
      iv: new Uint8Array(base64ToArrayBuffer(iv)),
    },
    importedKey,
    new Uint8Array(base64ToArrayBuffer(encryptedData)),
  );
  return new TextDecoder().decode(decryptedData);
}

export function isValidEncryptedDataMigrateKey(data: string): boolean {
  const { keyType, keyHash, iv, cipherText, algorithm } = JSON.parse(data);
  const sha256 = new Uint8Array(base64ToArrayBuffer(keyHash));
  if (keyType !== "migrateKey") {
    return false;
  }
  if (sha256.length !== 32) {
    return false;
  }
  if (new Uint8Array(base64ToArrayBuffer(iv)).length !== 12) {
    return false;
  }
  if (new Uint8Array(base64ToArrayBuffer(cipherText)).length !== 1088) {
    return false;
  }
  return true;
}

export function generateMigrateSignKey(): {
  publickKey: string;
  privateKey: string;
} {
  const seed = crypto.getRandomValues(new Uint8Array(32));
  const key = ml_dsa65.keygen(seed);
  const publicKeyBinary = arrayBufferToBase64(
    key.publicKey as unknown as ArrayBuffer,
  );
  const privateKeyBinary = arrayBufferToBase64(
    key.secretKey as unknown as ArrayBuffer,
  );
  const publickKey: migrateSignKey = {
    keyType: "migrateSignKeyPublic",
    key: publicKeyBinary,
  };
  const privateKey: migrateSignKey = {
    keyType: "migrateSignKeyPrivate",
    key: privateKeyBinary,
  };
  return {
    publickKey: JSON.stringify(publickKey),
    privateKey: JSON.stringify(privateKey),
  };
}

export function isValidMigrateSignKeyPrivate(key: string): boolean {
  if (key.length !== 5420) {
    console.log(key.length);
    return false;
  }
  const { key: keyBinary, keyType } = JSON.parse(key);
  if (keyType !== "migrateSignKeyPrivate") {
    return false;
  }
  const keyBinaryArray = new Uint8Array(base64ToArrayBuffer(keyBinary));
  if (keyBinaryArray.length !== 4032) {
    console.log(keyBinaryArray.length);
    return false;
  }
  return true;
}

export function isValidMigrateSignKeyPublic(key: string): boolean {
  if (key.length !== 2647) {
    console.log(key.length);
    return false;
  }
  const { key: keyBinary, keyType } = JSON.parse(key);
  if (keyType !== "migrateSignKeyPublic") {
    return false;
  }
  const keyBinaryArray = new Uint8Array(base64ToArrayBuffer(keyBinary));
  if (keyBinaryArray.length !== 1952) {
    console.log(keyBinaryArray.length);
    return false;
  }
  return true;
}

export function signDataMigrateSignKey(
  key: string,
  data: string,
  pubKeyHash: string,
): string | null {
  if (!isValidMigrateSignKeyPrivate(key)) {
    return null;
  }
  const { key: keyBinary } = JSON.parse(key);
  const dataArray = new TextEncoder().encode(data);
  const signature = ml_dsa65.sign(
    new Uint8Array(base64ToArrayBuffer(keyBinary)),
    dataArray,
  );
  const signString = arrayBufferToBase64(signature as unknown as ArrayBuffer);
  const signResult: Sign = {
    signature: signString,
    keyHash: pubKeyHash,
    keyType: "migrateSignKey",
    algorithm: "ML-DSA-65",
  };
  return JSON.stringify(signResult);
}

export function verifyDataMigrateSignKey(
  key: string,
  sign: string,
  data: string,
): boolean {
  if (!isValidMigrateSignKeyPublic(key)) {
    return false;
  }
  const { key: keyBinary } = JSON.parse(key);
  const signData: Sign = JSON.parse(sign);
  if (signData.keyType !== "migrateSignKey") {
    return false;
  }
  const dataArray = new TextEncoder().encode(data);
  const verify = ml_dsa65.verify(
    new Uint8Array(base64ToArrayBuffer(keyBinary)),
    dataArray,
    new Uint8Array(base64ToArrayBuffer(signData.signature)),
  );
  return verify;
}

export function isValidSignMigrateSignKey(sign: string): boolean {
  const { keyHash, signature, keyType } = JSON.parse(sign);
  if (sign.length !== 4512) {
    console.log(sign.length);
    return false;
  }
  if (keyType !== "migrateSignKey") {
    return false;
  }
  if (keyHash.length !== 44) {
    console.log(keyHash.length);
    return false;
  }
  if (signature.length !== 4412) {
    console.log(signature.length);
    return false;
  }
  return true;
}

File: rust/src/migrate_key.rs
use crate::r#type::{MigrateKey, MigrateSignKey, EncryptedData, Sign};
use crate::keyutils::{generate_kem_key_pair, generate_dsa65_key_pair};
use crate::crypto::{encrypt, decrypt};
use crate::utils::key_hash;
use crate::signature::{create_signature_object_mlds65, verify_with_mlds65};
use base64::engine::general_purpose::STANDARD as BASE64;
use base64::Engine as _;
use serde_json;

/// MigrateKey 生成
pub fn generate_migrate_key() -> (String,String) {
    let (pub_b64, priv_b64)=generate_kem_key_pair().unwrap();
    let pk=MigrateKey{key_type:"migrateKeyPublic".into(),key:pub_b64, timestamp:None};
    let sk=MigrateKey{key_type:"migrateKeyPrivate".into(),key:priv_b64,timestamp:None};
    (serde_json::to_string(&pk).unwrap(),serde_json::to_string(&sk).unwrap())
}
pub fn is_valid_migrate_key_public(json:&str)->bool {
    serde_json::from_str::<MigrateKey>(json).map(|k|k.key_type=="migrateKeyPublic"&&BASE64.decode(&k.key).map_or(false,|b|b.len()==1184)).unwrap_or(false)
}
pub fn is_valid_migrate_key_private(json:&str)->bool {
    serde_json::from_str::<MigrateKey>(json).map(|k|k.key_type=="migrateKeyPrivate"&&BASE64.decode(&k.key).map_or(false,|b|b.len()==2400)).unwrap_or(false)
}
pub fn encrypt_data_migrate_key(pub_json:&str, data:&str)->Option<String> {
    let mk: MigrateKey = match serde_json::from_str(pub_json) { Ok(v) => v, Err(_) => return None };
    if mk.key_type!="migrateKeyPublic" {return None}
    let enc = encrypt(data, &mk.key);
    let ed=EncryptedData{ key_type:"migrateKey".into(), key_hash:key_hash(pub_json), encrypted_data:enc.encrypted_data, iv:enc.iv, algorithm:Some(enc.algorithm), cipher_text:Some(enc.cipher_text) };
    serde_json::to_string(&ed).ok()
}
pub fn decrypt_data_migrate_key(priv_json:&str, json:&str)->Option<String> {
    let mk: MigrateKey=serde_json::from_str(priv_json).ok()?;
    if mk.key_type!="migrateKeyPrivate"{return None}
    let ed:EncryptedData=serde_json::from_str(json).ok()?;
    let ciphertext = ed.cipher_text.as_ref()?;
    Some(decrypt(&ed.encrypted_data, ciphertext, &ed.iv, &mk.key))
}
pub fn is_valid_encrypted_data_migrate_key(json:&str)->bool {
    serde_json::from_str::<EncryptedData>(json).map(|ed|ed.key_type=="migrateKey").unwrap_or(false)
}

/// MigrateSignKey 生成／署名／検証
pub fn generate_migrate_sign_key()->(String,String) {
    let (pub_b64, priv_b64)=generate_dsa65_key_pair().unwrap();
    let pk=MigrateSignKey{key_type:"migrateSignKeyPublic".into(),key:pub_b64,timestamp:None};
    let sk=MigrateSignKey{key_type:"migrateSignKeyPrivate".into(),key:priv_b64,timestamp:None};
    (serde_json::to_string(&pk).unwrap(),serde_json::to_string(&sk).unwrap())
}
pub fn sign_data_migrate_sign_key(priv_json:&str, data:&str, key_hash:&str)->Option<String> {
    let sk:MigrateSignKey=serde_json::from_str(priv_json).ok()?;
    if sk.key_type!="migrateSignKeyPrivate"{return None}
    create_signature_object_mlds65(&sk.key,data.as_bytes(),key_hash,"migrateSignKey").ok()
}
pub fn verify_data_migrate_sign_key(pub_json:&str, sign_json:&str, data:&str)->bool {
    let pk = match serde_json::from_str::<MigrateSignKey>(pub_json) {
        Ok(v) => v,
        Err(_) => return false,
    };
    let s = match serde_json::from_str::<Sign>(sign_json) {
        Ok(v) => v,
        Err(_) => return false,
    };
    if s.key_type != "migrateSignKey" { return false; }
    verify_with_mlds65(&pk.key, data.as_bytes(), &s.signature)
}
pub fn is_valid_sign_migrate_sign_key(json:&str)->bool {
    serde_json::from_str::<Sign>(json).map(|s|s.key_type=="migrateSignKey").unwrap_or(false)
}
